pgmq-extension/sql/pgmq--1.5.2--1.6.0.sql

```diff
@@ -0,0 +1,357 @@
+BEGIN
+    -- Create GIN index on headers for efficient FIFO key lookups
+    EXECUTE FORMAT(
+        $QUERY$
+        CREATE INDEX IF NOT EXISTS %I ON pgmq.%I USING GIN (headers);
+        $QUERY$,
+        index_name, qtable
+    );
+END;
+$$ LANGUAGE plpgsql;
+
+-- Helper function to create FIFO indexes on existing queues
+CREATE FUNCTION pgmq.create_fifo_index(queue_name TEXT)
+RETURNS void AS $$
+BEGIN
+    PERFORM pgmq._create_fifo_index_if_not_exists(queue_name);
+END;
+$$ LANGUAGE plpgsql;
+
+-- Helper function to create FIFO indexes on all existing queues
+CREATE FUNCTION pgmq.create_fifo_indexes_all()
+RETURNS void AS $$
+DECLARE
+    queue_record RECORD;
+BEGIN
+    FOR queue_record IN SELECT queue_name FROM pgmq.meta LOOP
+        PERFORM pgmq.create_fifo_index(queue_record.queue_name);
+    END LOOP;
+END;
+$$ LANGUAGE plpgsql;
+
+-- Create read_fifo_sqs_style function that mimics AWS SQS FIFO batch retrieval behavior
+-- This function attempts to return as many messages as possible from the same message group
+CREATE FUNCTION pgmq.read_fifo_sqs_style(
+    queue_name TEXT,
+    vt INTEGER,
+    qty INTEGER,
+    conditional JSONB DEFAULT '{}'
+)
+RETURNS SETOF pgmq.message_record AS $$
+DECLARE
+    sql TEXT;
+    qtable TEXT := pgmq.format_table_name(queue_name, 'q');
+BEGIN
+    sql := FORMAT(
+        $QUERY$
+        WITH fifo_groups AS (
+            -- Get the minimum msg_id for each FIFO group that's ready to be processed
+            SELECT 
+                COALESCE(headers->>'x-pgmq-fifo', '_default_fifo_group') as fifo_key,
+                MIN(msg_id) as min_msg_id
+            FROM pgmq.%I
+            WHERE vt <= clock_timestamp() 
+            AND CASE
+                WHEN %L != '{}'::jsonb THEN (message @> %2$L)::integer
+                ELSE 1
+            END = 1
+            GROUP BY COALESCE(headers->>'x-pgmq-fifo', '_default_fifo_group')
+        ),
+        group_priorities AS (
+            -- Assign priority to groups based on their oldest message
+            SELECT 
+                fifo_key,
+                min_msg_id,
+                ROW_NUMBER() OVER (ORDER BY min_msg_id) as group_priority
+            FROM fifo_groups
+        ),
+        available_messages AS (
+            -- Get messages prioritizing filling batch from earliest group first
+            SELECT 
+                m.msg_id,
+                gp.group_priority,
+                ROW_NUMBER() OVER (PARTITION BY gp.fifo_key ORDER BY m.msg_id) as msg_rank_in_group
+            FROM pgmq.%I m
+            INNER JOIN group_priorities gp ON 
+                COALESCE(m.headers->>'x-pgmq-fifo', '_default_fifo_group') = gp.fifo_key
+            WHERE m.vt <= clock_timestamp()
+            AND CASE
+                WHEN %L != '{}'::jsonb THEN (m.message @> %4$L)::integer
+                ELSE 1
+            END = 1
+            AND m.msg_id >= gp.min_msg_id  -- Only messages from min_msg_id onwards in each group
+        ),
```

Comment:
@v0idpwn , @tazmaniax  -- what if we did something like this?



```suggestion
            AND m.msg_id >= gp.min_msg_id  -- Only messages from min_msg_id onwards in each group
            AND NOT EXISTS (
                -- also ceck if any message in this group is currently being processed
                -- basically filter out messages currently invisible
               SELECT 1 
                FROM pgmq.%I m2 
                WHERE COALESCE(m2.headers->>'x-pgmq-fifo', '_default_fifo_group') = gp.fifo_key
                AND m2.vt > clock_timestamp()
              )
            FOR UPDATE SKIP LOCKED
        ),
```